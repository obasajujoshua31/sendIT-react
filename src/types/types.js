export const signUpUser = "SIGN_UP_USER";
export const signInUser = "SIGN_IN_USER";
export const createOrder = "CREATE_ORDER";
export const createOrderSuccess = "CREATE_ORDER_SUCCESS";
export const createOrderFailure = "CREATE_ORDER_FAILURE";
export const viewOrders = "VIEW_ORDERS";
export const viewOrder = "VIEW_ORDER";
export const updateParcel = "UPDATE_PARCELS";
export const updateLocation = "UPDATE_PARCEL_LOCATION";
export const cancelOrder = "CANCEL_ORDER";
export const updateParcelStatus = "UPDATE_PARCEL_STATUS";
export const signInSuccess = "SIGN_IN_SUCCESS";
export const signInFailure = "SIGN_IN_FAILURE";
export const signUpSuccess = "SIGN_UP_SUCCESS";
export const signUpFailure = "SIGN_UP_FAILURE";
export const isLoggedOut = "LOGGED_OUT";
export const isAdmin = "IS_ADMIN";
