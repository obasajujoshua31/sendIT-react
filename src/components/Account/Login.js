import React, { Component } from "react";

class Login extends Component {
  render() {
    return (
      <React.Fragment>
          <div className ='account-wrapper'>
        <div className="form-wrapper sign-wrapper">
        <h2>Sign In</h2>
          <div className="form-group">
            <label htmlFor="email">Email</label>
            <input type="email" placeholder ='yourname@example.com' />
          </div>
          <div className="form-group">
            <label htmlFor="password">Password</label>
            <input type="password" placeholder = 'your password' />
          </div>
          <button type = 'submit'>Login</button>
        </div>
        </div>
      </React.Fragment>
    );
  }
}

export default Login;