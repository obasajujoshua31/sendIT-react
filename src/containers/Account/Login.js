import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import propTypes from "prop-types";
import { toast } from "react-toastify";
import signUserIn from "../../actions/user/signin";
import { signInSuccess, signInFailure } from "../../types/types";
import { withRouter } from "react-router-dom";
// import Spinner from "react-spinkit";
import isValidLoginForm from "../../validations/loginValidator";

export class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: "",
      password: "",
      isLoading: false,
      isShowSpinner: false,
      formErrors: ""
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.singInStatus === signInSuccess) {
      if (nextProps.role) {
        this.props.history.push("/admin-dashboard/");
      } else {
        this.props.history.push("/view-orders/");
      }
      toast.success("You are successfully logged in");
    }
    if (nextProps.errorMessage) {
      toast.error(nextProps.errorMessage);
    }
  }
  handleChange(e) {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  }
  handleSubmit(e) {
    e.preventDefault();
    const { email, password } = this.state;
    if (!isValidLoginForm(this.state)) {
      return this.setState({
        formErrors: "Invalid Credentials!"
      });
    }
    // this.ShowSpinner();
    this.setState({ email: "", password: "" });
    this.props.signIn({ email, password });
  }

  render() {
    const { email, password, isShowSpinner, formErrors } = this.state;
    // const isflyingSpinner = <Spinner name="three-bounce" color="orange" />;

    return (
      <Fragment>
        <div className="account-wrapper">
          <div className="form-wrapper sign-wrapper">
            <form onSubmit={this.handleSubmit} noValidate>
              {formErrors.length > 0 && (
                <span className="errorMsg">{formErrors}</span>
              )}
              {/* {isShowSpinner ? isflyingSpinner : ""} */}
              <h2>Sign In</h2>
              {this.props.errorMessage.length > 0 && (
                <span className="errorMsg">{this.props.errorMessage}</span>
              )}
              <div className="form-group">
                <label htmlFor="email">Email</label>
                <input
                  type="email"
                  placeholder="yourname@example.com"
                  onChange={this.handleChange}
                  value={email}
                  name="email"
                />
              </div>
              <div className="form-group">
                <label htmlFor="password">Password</label>
                <input
                  type="password"
                  placeholder="your password"
                  onChange={this.handleChange}
                  value={password}
                  name="password"
                />
              </div>
              <button type="submit">Login</button>
            </form>
          </div>
        </div>
      </Fragment>
    );
  }
}
export const mapStateToProps = ({ users }) => ({
  singInStatus: users.singInStatus,
  errorMessage: users.errorMessage,
  role: users.role,
  errMsg: users.errMsg
});

export const mapDispatchToProps = dispatch => {
  return {
    signIn: user => dispatch(signUserIn(user))
  };
};

Login.propTypes = {
  signIn: propTypes.func.isRequired
};
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(Login));
