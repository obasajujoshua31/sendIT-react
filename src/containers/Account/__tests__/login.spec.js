import React from "react";
import { shallow, mount } from "enzyme";
import sinon from "sinon";
import { toast } from "react-toastify";
import { Login, mapDispatchToProps, mapStateToProps } from "../Login";
import { signUpSuccess, signInSuccess } from "../../../types/types";

const props = {
  signIn: jest.fn(),
  history: {
    push: jest.fn()
  },
  errorMessage: "",
  singInStatus: ""
};
const mockUser = {
  email: "monkey@gorilla.com",
  password: "animal"
};

// const mockToast = {
//     toast: jest.fn()
// }
const mockStore = {
  users: {
    singInStatus: "Success",
    errorMsg: null
  }
};

const wrapper = shallow(<Login {...props} />);

describe(" <Login> ", () => {
  it("should render Login Component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should contain 4 <div> tags", () => {
    expect(wrapper.find("div").length).toEqual(4);
  });
  it("should contain 1 <form> tag", () => {
    expect(wrapper.find("form").length).toEqual(1);
  });
  it("should contain 1 <h2> tag", () => {
    expect(wrapper.find("h2").length).toEqual(1);
  });

  it("should set the firstName field to test", () => {
    wrapper
      .find("input")
      .at(0)
      .simulate("change", {
        target: {
          value: "test",
          name: "firstName"
        }
      });
    expect(wrapper.instance().state.firstName).toEqual("test");
  });

  it("should call dispatch action", () => {
    const dispatchSpy = sinon.spy();
    const dispatchResult = mapDispatchToProps(dispatchSpy);
    dispatchResult.signIn(mockUser);
    expect(dispatchSpy.callCount).toEqual(1);
  });

  it("should map state to Props", () => {
    const mapStateResult = mapStateToProps(mockStore);
    expect(mapStateResult.singInStatus).toEqual("Success");
  });

  //   it("should call history.push", () => {
  //     const addedProps = { ...props, signUpStatus: signUpSuccess };
  //     const alteredWrapper = shallow(<Login {...addedProps} />);
  //     expect(alteredWrapper.instance().props.history.push).toHaveBeenCalled();
  //   });

  it("should call submit", () => {
    let prevented = false;
    const formElement = wrapper.find("form");
    formElement.simulate("submit", {
      preventDefault: () => {
        prevented = true;
      }
    });
    expect(prevented).toEqual(true);
  });

  it("should call props.signIn", () => {
    let prevented = false;
    wrapper.setState({
      email: mockUser.email,
      password: mockUser.password
    });
    const formElement = wrapper.find("form");
    formElement.simulate("submit", {
      preventDefault: () => {
        prevented = true;
      }
    });
    expect(wrapper.instance().props.signIn).toHaveBeenCalled();
  });

  it("should call history.push", () => {
    const addedProps = { ...props, singInStatus: signInSuccess };
    const alteredWrapper = shallow(<Login {...addedProps} />);
    alteredWrapper.setProps({
      role: true
    });
    expect(alteredWrapper.instance().props.history.push).toHaveBeenCalled();
  });
  it("should call history.push", () => {
    const addedProps = { ...props, singInStatus: signInSuccess };
    const alteredWrapper = shallow(<Login {...addedProps} />);
    alteredWrapper.setProps({
      role: false
    });
    expect(alteredWrapper.instance().props.history.push).toHaveBeenCalled();
  });
  it("should call history.push", () => {
    // const addedProps = { ...props };
    const alteredWrapper = shallow(<Login {...props} />);
    alteredWrapper.setProps({
      errorMessage: "Error"
    });
    
    // const toastSpy = jest.spyOn(toast, 'error')
    // // const toastSpy = toast(warn)
    // // const toastSpy = toast(mockToast);
    // expect(toastSpy).toHaveBeenCalledWith('Error');
  });
});
