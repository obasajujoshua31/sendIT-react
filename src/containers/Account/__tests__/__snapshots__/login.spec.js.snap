// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` <Login>  should render Login Component 1`] = `
<Fragment>
  <div
    className="account-wrapper"
  >
    <div
      className="form-wrapper sign-wrapper"
    >
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <h2>
          Sign In
        </h2>
        <div
          className="form-group"
        >
          <label
            htmlFor="email"
          >
            Email
          </label>
          <input
            name="email"
            onChange={[Function]}
            placeholder="yourname@example.com"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="password"
          >
            Password
          </label>
          <input
            name="password"
            onChange={[Function]}
            placeholder="your password"
            type="password"
            value=""
          />
        </div>
        <button
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>
</Fragment>
`;
